onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -237,6 +406,80 @@ def calculate_quantization_params(self, quantization_thresholds):
+                   black_nodes=[],
+                   white_nodes=[],
+                   augmented_model_path='augmented_model.onnx'):
+
+    calibrator = ONNXCalibrater(model_path, data_reader, op_types, black_nodes, white_nodes, augmented_model_path)
+
+    return calibrator
+
+def calculate_calibration_data(model_path,
+                               calibrator=None,
+                               calibration_data_reader: CalibrationDataReader=None,
+                               op_types_to_quantize=[],
+                               activation_type=QuantType.QUInt8,
+                               nodes_to_quantize=[],
+                               nodes_to_exclude=[],
+                               augmented_model_path='augmented_model.onnx'):
+
+    if activation_type != QuantType.QUInt8:
+        raise ValueError("Static quantization only support uint8 for activation now.")
+
+    if not op_types_to_quantize or len(op_types_to_quantize) == 0:
+        op_types_to_quantize = list(QLinearOpsRegistry.keys())
+
+    print("model path: %s" % model_path)
+    print("augmented model path: %s" % augmented_model_path)
+
+    if not calibrator:
+        calibrator = get_calibrator(model_path, calibration_data_reader, op_types_to_quantize, nodes_to_quantize, nodes_to_exclude, augmented_model_path=augmented_model_path)
+
+    if not os.path.exists(augmented_model_path):
+        augmented_model = calibrator.augment_graph(augment_all_ops=True)
+        onnx.save(augmented_model, augmented_model_path)
+
+    calibrator.get_intermediate_outputs(providers=["CUDAExecutionProvider"])
+
+def generate_calibration_table(model_path, augmented_model_path, data_reader, calibration_dataset=None, stride=5000, batch_size=20):
+
+    if os.path.exists(augmented_model_path):
+        os.remove(augmented_model_path)
+        print("remove previously generated %s and start to generate a new one." % (augmented_model_path))
+
+
+    calibrator = None 
+
+    # If no dataset is provided upfront, the dataset will be fetched by data reader itself and 
+    # therefore no need to handle batch processing at this moment. Just make the loop to run only once.
+    total_data_size = len(os.listdir(calibration_dataset)) if calibration_dataset else stride
+
+
+    # Some machines don't have sufficient memory to hold all dataset at once. So handle it by batch/stride.
+    # For each stride, data_reader can handle them with batch or serial processing depends on data reader implementation 
+    for i in range(0, total_data_size, stride):
+        print(calibration_dataset)
+        if calibration_dataset:
+            print("Total data size %s\nStart to process data from index %s with stride %s ..." % (str(total_data_size), str(i), str(stride)))
+            data_reader.set_start_index(i)
+            data_reader.set_size_limit(stride)
+            data_reader.set_batch_size(batch_size)
+            data_reader.set_preprocess_flag(True)
```

Comment:
all those are not common interface of DataReader

Reply:
The reason to have these functions is that I want to make existed data_reader instance capable of setting new configuration to pre-process new input data without newing another data_reader instance.
For this scenario, the dataset has 40000 images and current T4 machine's memory can't hold all the images at once. So we have to do it with stride, as generate_calibration_table() does here. 
generate_calibration_table() has one argement, data_reader, and this function doesn't know what kind of data_reader it is.
So it can't new the data_reader, but I can set the data_reader to preprocess new stride of image data.

Reply:
This logic should be handled in the DataReader. The data reader should keep the status of reading and be implemented as something like a streaming. get_next read a batch of images, and then next call of get_next remove the last batch and load another batch. You don't need to load all the 40k images at once, right?

---
In reply to: [541832012](https://github.com/microsoft/onnxruntime/pull/6103#discussion_r541832012) [](ancestors = 541832012)

Reply:
I've move this logic to DataReader. As you suggest, data reader keep the status of reading and get_next() read a batch of images until all the batches in the dataset are being read.